src/compiler/factory/base_node_factory.rs

```diff
@@ -1,8 +1,68 @@
-use crate::{BaseNode, SyntaxKind};
+use std::cell::RefCell;
+
+use crate::{object_allocator, BaseNode, SyntaxKind};
 
 pub trait BaseNodeFactory {
     fn create_base_source_file_node(&self, kind: SyntaxKind) -> BaseNode;
     fn create_base_identifier_node(&self, kind: SyntaxKind) -> BaseNode;
     fn create_base_token_node(&self, kind: SyntaxKind) -> BaseNode;
     fn create_base_node(&self, kind: SyntaxKind) -> BaseNode;
 }
+
+pub fn create_base_node_factory() -> BaseNodeFactoryConcrete {
+    BaseNodeFactoryConcrete::new()
+}
+
+#[allow(non_snake_case)]
+pub struct BaseNodeFactoryConcrete {
+    SourceFileConstructor: RefCell<Option<fn(SyntaxKind, isize, isize) -> BaseNode>>,
+    IdentifierConstructor: RefCell<Option<fn(SyntaxKind, isize, isize) -> BaseNode>>,
+    TokenConstructor: RefCell<Option<fn(SyntaxKind, isize, isize) -> BaseNode>>,
+    NodeConstructor: RefCell<Option<fn(SyntaxKind, isize, isize) -> BaseNode>>,
+}
+
+impl BaseNodeFactoryConcrete {
+    pub fn new() -> Self {
+        Self {
+            SourceFileConstructor: RefCell::new(None),
+            IdentifierConstructor: RefCell::new(None),
+            TokenConstructor: RefCell::new(None),
+            NodeConstructor: RefCell::new(None),
+        }
+    }
+}
+
+#[allow(non_snake_case)]
+impl BaseNodeFactory for BaseNodeFactoryConcrete {
+    fn create_base_source_file_node(&self, kind: SyntaxKind) -> BaseNode {
+        let mut SourceFileConstructor = self.SourceFileConstructor.borrow_mut();
+        if SourceFileConstructor.is_none() {
+            *SourceFileConstructor = Some(object_allocator.get_source_file_constructor());
+        }
+        (SourceFileConstructor.unwrap())(kind, -1, -1)
```

Comment:
Combined with the if-statement above, this seems equivalent to `unwrap_or`?

Reply:
Would that accomplish the mutation of `self.SourceFileConstructor`?

Reply:
Oh I see. No if the mutation needs to persist beyond this function that doesn't work.